### DISCUSSION TITLE
how to set a prefix for a code block (prefix rendered, but not copy-able by clipboardjs)

### DISCUSSION THREAD

**Discussion by flynneva** [squidfunk/mkdocs-material#3427] [Q&A] [answered]
I'd like to be able to render a prefix for lines in a code block to indicate the environment (docker, venv, etc.) but would like this prefix to _not be copyable_ by the "copy to clipboard" button.

Is this already possible today with some configuration?

**Comment by squidfunk** [comment]
You probably mean something like `$` to designate a command prompt. In short: no, it's not possible to do in a general way, because it's undecidable whether a line should be prefixed or not. However, you might be able to prepend something to each line with a pseudo-element and some CSS.

**Reply by flynneva** [reply]
Awesome thanks for the tips @squidfunk! Any chance there is already some documentation somewhere that might help me figure out the CSS + pseudo-element proposal you gave?

**Reply by squidfunk** [reply]
None that I know of, but it was discussed somewhere before, probably over at the MkDocs issue tracker or discussion board.

**Comment by flynneva** [comment, marked as answer]
as a possible feature request, do you think it would make sense / be possible to edit [the `linenumber` feature](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/?h=code#adding-line-numbers) of code blocks and instead add arbitrary strings instead?

Something like:

````
```py linenums="(venv)$"
def bubble_sort(items):
    for i in range(len(items)):
        for j in range(len(items) - 1 - i):
            if items[j] > items[j + 1]:
                items[j], items[j + 1] = items[j + 1], items[j]
```
````

Which renders to:

```py linenums="(venv)$"
(venv)$ def bubble_sort(items):
(venv)$    for i in range(len(items)):
(venv)$        for j in range(len(items) - 1 - i):
(venv)$            if items[j] > items[j + 1]:
(venv)$                items[j], items[j + 1] = items[j + 1], items[j]
```

but the `(venv)$` is not copied when you click the `copy to clipboard` button

**Reply by squidfunk** [reply]
It's currently not possible to pass arbitrary attributes via code blocks. This is a limitation of either Python Markdown or Python Markdown Extensions which is an upstream dependency of ours. Other than that I don't think the `linenums` attribute supports this.

**Reply by facelessuser** [reply]
You can pass arbitrary `data-attr` to the code block, but the way CSS works, the `data-attr `will not be seen by the linenum spans, only the element on which the attribute is attached.

With that said, it is not impossible to do. You could post-process such blocks with JavaScript or you could post-process them with a [custom fence](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/#custom-fences). While I have no intention of coding up a complete custom solution for anyone, I may demonstrate a general fence that could be adapted to do what you require, mainly a fence that wraps the default highlight logic.  I imagine it is possible to wrap the SuperFence highlight logic and then post process the returned code blocks extracting a `data-attr` with your desired prefix and then modifying the line numbers such that they will render your prefix instead of numbers.

**Reply by squidfunk** [reply]
Ah, I thought only `title` is supported. I'm sorry, I may have misread your statement in our last correspondence on code blocks 😅

**Reply by facelessuser** [reply]
Yeah, I also implemented  `data-attr`. I guess maybe I wasn't clear. `data-attr` are only available in the scope of the element in which it is applied, not children though, so this scenario is not easily handled with CSS alone. I think a custom fence or JS could do the trick though.

**Reply by flynneva** [reply]
@facelessuser wow thanks for the info - this seems to be exactly what I am looking for. From your advice and the [existing mkdocs-material docs on superfences](https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown-extensions/#superfences) I think I should be able to figure this out on my own.

Thanks much, and as always awesome stuff here - I love working with mkdocs-material! 😄

**Reply by bachirelkhoury** [reply]
> @facelessuser this seems to be exactly what I am looking for. From your advice and the [existing mkdocs-material docs on superfences](https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown-extensions/#superfences) I think I should be able to figure this out on my own.


@flynneva I have a similar use case... looking for a similar solution.. did you manage to get it worked?
interested to see how did you go about solving this.

thanks

**Reply by flynneva** [reply]
hey @bachirelkhoury apologies for the delay on following up here - just catching up with my github pings now.

Unfortunately I haven't gotten to this task yet but I will follow up here once I do figure it out, don't worry and I'll be sure to ping you as well.

**Reply by flynneva** [reply]
@bachirelkhoury @squidfunk @facelessuser I finally got around to implementing this and I think we ended up with a super flexible solution. Posting here for anyone else who is interested.

1. add a new custom superfence to your `mkdocs.yml` and specify the format function:

```
- pymdownx.superfences:
    custom_fences:
        - name: prefix
          class: prefix
          format: !!python/name:custom_doc_tools.superfences.prefix_code_format
```

2. Implement the format function in a python module somewhere. Within this formatter function, trick the normal superfences highlight formatter to do the hard work for you:


```
from bs4 import BeautifulSoup

from pymdownx.superfences import SuperFencesBlockPreprocessor, SuperFencesCodeExtension

def prefix_code_format(source, language, class_name, options, md, **kwargs):
       num_lines = len(source.splitlines())
        if 'prefix' in options:
            prefix = options['prefix']
        else:
            prefix = ''

        if 'classes' in kwargs:
            classes = kwargs['classes']
        else:
            classes = []

        if options['lines'] == 0:
            # assume prefix to be set for every line in codeblock
            options['lines'] = f"1-{num_lines}"
        lines = options['lines']
        # handle the case where prefix and hl_lines are specified
        keep_hl = False
        if 'hl_lines' in options:
            keep_hl = True
        if 'hl_lines' not in classes:
            classes.append('hl_lines')
        # set hl_lines to lines to reuse hl_lines formatter logic
        options['hl_lines'] = lines
        # use linenum preprocessor to replace them later with prefix
        if 'linenums' not in options:
            options['linenums'] = '1'

        sf_ext = SuperFencesCodeExtension()
        config = sf_ext.getConfigs()
        preprocessor = SuperFencesBlockPreprocessor(md)
        preprocessor.config = config
        preprocessor.extension = sf_ext
        preprocessor.get_hl_settings()
        preprocessor.line_count = num_lines
        lines_int = preprocessor.parse_hl_lines(lines)
        soup = BeautifulSoup(
            preprocessor.highlight(
                src=source,
                language=language,
                options=options,
                md=md,
                classes=classes,
                id_value='',
                attrs=None),
            features='lxml')
        line_nums_col = soup.find("div", {"class": "linenodiv"})
        for index, row in enumerate(line_nums_col.findAll(class_='normal')):
            if (index + 1) in lines_int:  # index starts at 0, lines expected to start at 1
                row.string = prefix
            else:
                row.string = ''
```

3. Use in your `.md` files like this:

~~~
```prefix prefix="(venv)$" lines="2-3"
source venv/bin/activate
echo "I am in a venv"
deactivate
echo "I am not in a venv"
```
~~~

And you should see something like this:

![image](https://user-images.githubusercontent.com/6157095/182497915-aa89f3f3-bb2b-42c6-a4c6-07271d009ebe.png)

**Reply by mycroftsnm** [reply]
Hi, noob here;  i tried this and its not working :( , i'm not really sure if i installed correctly, just created a .py in the root of the project, and run with python3 -m mkdocs , but the example provided its not working, i also tried:

````
``` {.prefix linenums="1" prefix="test" lines="0"}
import foo.bar
```
````

in md file
but it renders literally like 
```{.prefix linenums="1" prefix="test" lines="0"} import foo.bar```


The example also renders like : 
```prefix prefix="(venv)$" lines="2-3" source venv/bin/activate echo "I am in a venv" deactivate echo "I am not in a venv"```


can u provide specific instructions about the installation ? thanks

**Reply by flynneva** [reply]
@mycroftsnm here is [a quick demo i just whipped up tonight](https://github.com/flynneva/mkdocs-custom-superfence-example). Check it out and let me know if that helps you out or if it is still missing something.

@squidfunk feel free to take any examples from that demo repo and move them into the mkdocs-material docs if desired.

**Reply by mycroftsnm** [reply]
Works perfect now, thanks :)

**Comment by ben519** [comment]
@facelessuser helped me (mostly) accomplish this, as discussed [here](https://github.com/facelessuser/pymdown-extensions/issues/1802).

**extra.css**

```
/* shell prompt */
.language-console .gp {
    font-weight: 800;
    user-select: none;
}
.language-console .go {
    font-style: italic;
}
```

**somepage.md**

    ```console
    bill@gates:~$ git log
    commit 88d8887beaad2cfd37d04fb00f38c05077581057 (HEAD -> main)
    Author: bill123 <billgates@gmail.com>
    Date:   Sun Sep 11 13:09:50 2022 -0500
    
        first commit
    ```

This produces output like [this](https://www.practiceprobs.com/problemsets/git/beginner/first-commit/solution/). 

<img width="703" alt="Screen Shot 2022-09-20 at 7 33 30 PM" src="https://user-images.githubusercontent.com/723099/191388512-5a2b5277-afc2-4aef-8e83-aadc7181a52d.png">


The copy button copies _everything_, but if you click and drag to highlight the code, it will not copy the prompt (`bill@gates:~$`).

**Comment by shepherdjerred** [comment]
You can do this with the following added TypeScript -- you'll need to compile before using. Define styles for `console-input` and `console-output`.

Example output:

<img width="711" alt="CleanShot 2023-04-04 at 20 57 58@2x" src="https://user-images.githubusercontent.com/3904778/229977557-a73b9821-28be-4f92-acb2-81764f4101e9.png">

Only commands are selectable:

<img width="704" alt="CleanShot 2023-04-04 at 20 58 22@2x" src="https://user-images.githubusercontent.com/3904778/229977592-ee3328f9-b83f-495b-924b-54026452541b.png">

Copying omits the comments, output, and prefix. This is what's copied:

```
rspm create source --name=local-python-src --type=local-python
```

```typescript
const codeBlocks = [...document.querySelectorAll(".highlight > pre > code")];

const rules = [
  // shell commands
  {
    prefix: "$ ",
    classes: ["code-noselect", "console-input"],
  },
  // comments
  {
    prefix: "# ",
    plainText: true,
    entireLine: true,
    classes: ["code-noselect", "c1"],
  },
  // standard output
  {
    prefix: "<< ",
    plainText: true,
    entireLine: true,
    deletePrefix: true,
    classes: ["code-noselect", "console-output"],
  },
];

const zip = (a, b) => a.map((k, i) => [k, b[i]]);

// update code blocks with the mappings defined above
codeBlocks.forEach((block) => {
  let text = (block.textContent || "").split("\n");
  let html = block.innerHTML.split("\n");
  let lines = zip(text, html);

  lines = lines.map((line) => {
    const lineText = line[0];
    const lineHtml = line[1];

    const matches = rules.filter((rule) => lineText.startsWith(rule.prefix));

    if (matches.length > 1) {
      throw new Error("multiple rule matches");
    }

    if (matches.length === 1) {
      const matchingRule = matches[0];
      const classes = matchingRule.classes.join(" ");
      let newContent: string;

      if (matchingRule.plainText) {
        newContent = lineText;
      } else {
        newContent = lineHtml;
        newContent = newContent.replace(
          matchingRule.prefix.replace(" ", `<span class="w"> </span>`),
          matchingRule.prefix
        );
      }

      if (matchingRule.entireLine) {
        newContent = `<span class="${classes}">${newContent}</span>`;
      } else {
        newContent = newContent.replace(
          matchingRule.prefix,
          `<span class="${classes}">${matchingRule.prefix}</span>`
        );
      }

      if (matchingRule.deletePrefix) {
        newContent = newContent.replace(matchingRule.prefix, "");
      }

      return newContent;
    } else {
      return lineHtml;
    }
  });

  block.innerHTML = lines.join("\n").trim();
});

// Update html attributes to remove elements with `code-noselect` when copying to clipboard
window.addEventListener("load", () => {
  const ids = document.querySelectorAll("*[data-clipboard-target]");
  ids.forEach((id) => {
    const attr = id.getAttribute("data-clipboard-target");
    if (!attr) {
      return;
    }
    const node = document.querySelector(attr);
    if (!node) {
      return;
    }
    let text: string[] = [];
    node.childNodes.forEach((child) => {
      if (child.nodeName === "#text") {
        text.push(child.textContent || "");
      } else if (!(child as any).classList.contains("code-noselect")) {
        text.push((child as any).innerText);
      }
    });
    let joinedText = text.join("");
    id.setAttribute(
      "data-clipboard-text",
      joinedText.trim().replace(/(?:\h*\n){2,}/g, "\n")
    );
    id.removeAttribute("data-clipboard-target");
  });
});

```
